File: src/domain/ids.ts
import { randomBytes } from "node:crypto";

import type { State } from "../types";

/**
 * Crockford base32 alphabet used by ULID.
 * 8 chars at 5 bits each = 40 bits of entropy (~1 trillion combinations).
 */
const CROCKFORD = "0123456789abcdefghjkmnpqrstvwxyz";

/**
 * Generate a root task ID using ULID short-form encoding.
 * Produces `tsq-<8 crockford base32 chars>` with 40 bits of randomness.
 *
 * Parameters are accepted for backward compatibility but ignored;
 * all IDs are now random to maximize entropy.
 */
export const makeRootId = (_title?: string, _nonce?: string): string => {
  const bytes = randomBytes(5); // 40 bits
  let id = "";
  let bits = 0;
  let acc = 0;
  for (const byte of bytes) {
    acc = (acc << 8) | byte;
    bits += 8;
    while (bits >= 5) {
      bits -= 5;
      id += CROCKFORD[(acc >> bits) & 0x1f];
    }
  }
  return `tsq-${id}`;
};

// child_counters[parentId] is maintained by the projector during task.created
// event application (setChildCounter), so it is always authoritative and we
// do not need to scan all tasks to find the max child index.
export const nextChildId = (state: State, parentId: string): string => {
  const maxChild = state.child_counters[parentId] ?? 0;
  return `${parentId}.${maxChild + 1}`;
};

File: src-rust/src/domain/ids.rs
use crate::types::State;
use rand::RngCore;
use rand::rngs::OsRng;

const CROCKFORD: &[u8; 32] = b"0123456789abcdefghjkmnpqrstvwxyz";

pub fn make_root_id(_title: Option<&str>, _nonce: Option<&str>) -> String {
    let mut bytes = [0u8; 5];
    OsRng.fill_bytes(&mut bytes);
    let mut id = String::with_capacity(8);
    let mut bits = 0u32;
    let mut acc = 0u64;
    for byte in bytes {
        acc = (acc << 8) | byte as u64;
        bits += 8;
        while bits >= 5 {
            bits -= 5;
            let index = ((acc >> bits) & 0x1f) as usize;
            id.push(CROCKFORD[index] as char);
        }
    }
    format!("tsq-{}", id)
}

pub fn next_child_id(state: &State, parent_id: &str) -> String {
    let max_child = state.child_counters.get(parent_id).copied().unwrap_or(0);
    format!("{}.{}", parent_id, max_child + 1)
}
